### PR TITLE
[RISCV] Add Zicfiss support to the shadow call stack implementation.

### PR DESCRIPTION
**Pull request by yetingk**
This patch enable hardware shadow stack with `Zicifss` and `mno-forced-sw-shadow-stack`. New feature forced-sw-shadow-stack disables hardware shadow stack even when `Zicfiss` enabled.

### CODE REVIEW

**Comment by samitolvanen** [comment]
The patch basically changes the ShadowCallStack back-end to emit an sspush/sspopchk instead of the usual SCS push/pop, which seems like a reasonable approach to me. However, it would be helpful to mention the dependency on `-fsanitize=shadow-call-stack` in the commit message, and you should also update [the documentation](https://clang.llvm.org/docs/ShadowCallStack.html).

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:115]: Do we need an enable other than just the feature being in -march? The shadow stack pointer has to be set up when the application starts.  Is this done by the kernel?

My concern is that if your -mcpu supports Zicfiss, but the kernel doesn't, this will generate code that doesn't work.

Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:115]: I think this is probably handled by the loader. Its how it is currently handled for platforms like Fuchsia and Android when SCS is enabled. I'm not totally sure on the new spec, but the linked document above makes me think that is the case with this text:

```
The dynamic loader should activate the use of Zicfiss extension for an application only if all executables (the application and the dependent dynamically linked libraries) used by that application use the Zicfiss extension.

An application that has the Zicfiss extension active may request the dynamic loader at runtime to load a new dynamic shared object (using dlopen() for example). If the requested object does not have the Zicfiss attribute then the dynamic loader, based on its policy (e.g, established by the operating system or the administrator) configuration, could either deny the request or deactivate the Zicfiss extension for the application. It is strongly recommended that the policy enforces a strict security posture and denies the request.
```

I believe that setting up this memory and initializing the SCS reg would fall under `activate the use of Zicfiss...`, but I could be wrong.

Comment (kito-cheng) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:115]: The plan is: adding GNU property and then check that during loader (and check at CRT stuff for static linked binary), that's mostly same as AArch64 and x86 for their CFI stuffs, and abort if system don't have either zicfiss and zimop, something begin implemented at our end around glibc, but we just didn't document everything down yet :P

Comment (kito-cheng) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:115]: > I believe that setting up this memory and initializing the SCS reg would fall under activate the use of Zicfiss..., but I could be wrong.

That suppose initialize SCS register by kernel, more precisely is user space will invoke some prctl system call to enable SCS and the kernel will allocate SCS page and then initialize the SCS register then back to user space.

Comment (kito-cheng) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:115]: > Do we need an enable other than just the feature being in -march? The shadow stack pointer has to be set up when the application starts. Is this done by the kernel?
>
> My concern is that if your -mcpu supports Zicfiss, but the kernel doesn't, this will generate code that doesn't work.

oh, that's good point...let me check how other target do

Comment (samitolvanen) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:115]: > > Do we need an enable other than just the feature being in -march? The shadow stack pointer has to be set up when the application starts. Is this done by the kernel?
> > My concern is that if your -mcpu supports Zicfiss, but the kernel doesn't, this will generate code that doesn't work.
> 
> oh, that's good point...let me check how other target do

If the kernel doesn't support Zicfiss, it presumably also hasn't enabled the extension for U-mode. Therefore, my understanding is that the application loader can simply skip shadow stack setup for the program, including the `prctl` (which would fail with `-EINVAL`), and since the Zicfiss (Zimop) instructions are no-ops when the extension is disabled, the program would run normally, just without shadow stacks. Does this sound correct?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:115]: > > Do we need an enable other than just the feature being in -march? The shadow stack pointer has to be set up when the application starts.
> 
> The hardware shadow stack support in this pr needs `-fsanitize=shadow-call-stack` now.

What if we're compiling for a platform that only uses the software shadow stack and does not support the hardware shadow stack even if the CPU supports it?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (ilovepi) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:30]: is it possible to use something other than a `cl::opt`? I know we make heavy use of them to control codegen options, but they tend to get dropped on the floor easily if you have to forward `-mllvm` options to the linker in LTO/ThinLTO builds. This is a constant pain point for a lot of folks who need to support multiple build configurations, like the Linux kernel, Fuchsia, Android, or Chrome.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: > What if we're compiling for a platform that only uses the software shadow stack and does not support the hardware shadow stack even if the CPU supports it?

Today, if you use software SCS on a platform without support they just fail, since the SCS won't be setup and gp(probably) points to invalid memory. Is it unreasonable to have it work the same in this case?

WDYT about phrasing this an an opt-out flag, like `-force-software-scs`, and using a helper function, like :
```
bool enableHWSCS(){ return !ForceSoftwareSCS && STI.hasFeature(RISCV::FeatureStdExtZicfiss); }
```
My thinking is that using HW SCS on platforms that have it is a good default, which we want to encourage, but we can still allow folks to force the software implementation if they want. 

It also occurs to me that X86 is getting/has backward edge CFI, right? how do we enable that in the x86 backend? How is GCC planning to spell these options, since I would suppose there needs to be some level of compatibility.

Comment (samitolvanen) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: > Today, if you use software SCS on a platform without support they just fail, since the SCS won't be setup and gp(probably) points to invalid memory. Is it unreasonable to have it work the same in this case?

Sounds reasonable to me, but having a flag that allows one to explicitly select between software and hardware shadow stacks is still a good idea IMO.

> WDYT about phrasing this an an opt-out flag, like `-force-software-scs`, and using a helper function, like :

For sanitizers, extra flags have typically been something like `-fsanitize-[sanitizer]-[flag]`, e.g. `-fsanitize-cfi-icall-generalize-pointers`:

https://clang.llvm.org/docs/ControlFlowIntegrity.html#indirect-function-call-checking

> It also occurs to me that X86 is getting/has backward edge CFI, right? how do we enable that in the x86 backend? How is GCC planning to spell these options, since I would suppose there needs to be some level of compatibility.

X86 uses the `-fcf-protection` flag for backward-edge at least, and AArch64 uses `-mbranch-protection` for both backward-edge (PAC) and forward-edge (BTI). It would be great not to invent another command line option specifically for RISC-V, so I quite like the idea of reusing the existing ShadowCallStack front-end for hardware shadow stacks too.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: > Today, if you use software SCS on a platform without support they just fail, since the SCS won't be setup and gp(probably) points to invalid memory. Is it unreasonable to have it work the same in this case?

If the hardware shadow stack isn't set up, won't the instructions become silent NOPs rather than failing?

Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: I thought that the scenario was that the platform only supported software, but the CPU did support the instructions, and would therefore execute `SSPUSH`/`SSPOPCHK`. I would expect those instructions fail when the memory region wasn't initialized and the `SSP` register would point to invalid memory.

If that's not what you meant, then I've misunderstood the scenario.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: As @samitolvanen wrote earlier the SSPUSH and SSPOPCHK instruction fall back to NOPs based on configuration.

> If the kernel doesn't support Zicfiss, it presumably also hasn't enabled the extension for U-mode. Therefore, my understanding is that the application loader can simply skip shadow stack setup for the program, including the prctl (which would fail with -EINVAL), and since the Zicfiss (Zimop) instructions are no-ops when the extension is disabled, the program would run normally, just without shadow stacks. Does this sound correct?

Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: Ah, you're totally right. I read that bit incorrectly. Thanks for pointing that out.

File: clang/docs/ShadowCallStack.rst
Comment (topperc) [inline clang/docs/ShadowCallStack.rst:62]: Was `-mllvm -riscv-hardware-shadow-stack` removed and replaced with `forced-sw-shadow-stack` feature?

**Comment by topperc** [comment]
How is unwinding handled? The software shadow stack documentation says this

> Special unwind information is required on functions that are compiled with ShadowCallStack and that may be unwound, i.e. functions compiled with -fexceptions (which is the default in C++). Some unwinders (such as the libgcc 4.9 unwinder) do not understand this unwind info and will segfault when encountering it. LLVM libunwind processes this unwind info correctly, however. This means that if exceptions are used together with ShadowCallStack, the program must use a compatible unwinder.

This patch skips the emission of the unwind info.

File: clang/docs/ShadowCallStack.rst
Comment (ilovepi) [inline clang/docs/ShadowCallStack.rst:64]:
```suggestion
hardware shadow stack, which needs `Zicfiss`_. Note that with ``Zicfiss``_ the RISC-V backend will default to the hardware based shadow call stack. Users can force the RISC-V backend to generate the software shadow call stack with ``Zicfiss``_ by passing ``-mforced-sw-shadow-stack``. 
```
Maybe this is easier to follow? Might also make sense to outline when `-mforced-sw-shadow-stack` is expected to work somewhere, e.g., it has no effect w/o `Zicfiss``.  TBH, in that case, the driver should probably issue an error, since its incompatible w/o the feature, but I'm not sure if we can do that check in the frontend or not.

File: clang/docs/ShadowCallStack.rst
Comment (ilovepi) [inline clang/docs/ShadowCallStack.rst:161]: We reserve `gp` even w/o SCS. The `--no-relax-gp`  advice does only apply to the software SCS.

Maybe something along these lines?

```suggestion
``-ffixed-x18`` unless your target already reserves ``x18``. No additional flags need to be passed on RISC-V because the software based shadow stack uses ``x3`` (``gp``), which is always reserved, and the hardware based shadow call stack uses a dedicated register, ``ssp``.
However, it is important to disable GP relaxation in the linker when using the software based shadow call stack on RISC-V. 
This can be done with the ``--no-relax-gp`` flag in GNU ld, and is off by default in LLD.
```

**Review by ilovepi — COMMENTED** [review]
Should the driver issue an error when using `-mforced-sw-shadow-stack` w/o `Zicfiss`? As mentioned in-line, I'm not sure we can do that check, but it feels like it should be incompatible.

**Comment by ilovepi** [comment]
The title should probably be changed to something like:

`[RISCV] Add Zicfiss support to the shadow call stack implementation`

The current title doesn't make much sense to me, but maybe I've misunderstood the intent?

File: llvm/lib/Target/RISCV/RISCVFeatures.td
Comment (ilovepi) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:1049]: Should the default be `false`?

File: clang/docs/ShadowCallStack.rst
Comment (ilovepi) [inline clang/docs/ShadowCallStack.rst:62]:
```suggestion
hardware shadow stack, which needs `Zicfiss`
```
I don't think that the default we want is `use SW shadow stack`, right? I think that using the SW based SCS should be something users opt into when `Zicfiss` is available, and shouldn't be something they have to opt out of...

File: clang/docs/ShadowCallStack.rst
Comment (ilovepi) [inline clang/docs/ShadowCallStack.rst:69]: Nit: this line is pretty long compared to the rest of the file.

File: clang/docs/ShadowCallStack.rst
Comment (ilovepi) [inline clang/docs/ShadowCallStack.rst:162]: Nit: this line is long. Maybe reformat the sentence/paragraph?

File: clang/test/Driver/riscv-features.c
Comment (ilovepi) [inline clang/test/Driver/riscv-features.c:30]: After commenting about defaults, we should probably test the default, too.

**Review by ilovepi — COMMENTED** [review]
I think this is basically good IMO once we settle on the default values. You should still get approval from one of the RISCV code owners before landing though.

File: clang/docs/ShadowCallStack.rst
Comment (topperc) [reply clang/docs/ShadowCallStack.rst:62]: Should Android default to shadow stack?

My concern is that if a user passes an -mcpu that supports Zicfiss, but the OS doesn't, the compiler will emit shadow stack instructions that silently fall back to NOPs due to the missing OS support. This leaves your binary in a state where it gets no protection.

Comment (samitolvanen) [reply clang/docs/ShadowCallStack.rst:62]: > Should Android default to shadow stack?
> 
> My concern is that if a user passes an -mcpu that supports Zicfiss, but the OS doesn't, the compiler will emit shadow stack instructions that silently fall back to NOPs due to the missing OS support. This leaves your binary in a state where it gets no protection.

I agree, changing `-fsanitize=shadow-call-stack` behavior based on `-mcpu` is problematic, especially when it can result in the program silently falling back to unprotected state. This might be a problem for other platforms too, not only Android.

Comment (topperc) [reply clang/docs/ShadowCallStack.rst:62]: @enh-google @appujee @samitolvanen @ilovepi What should the default behavior be. Changing based on -mcpu seems surprising to me.

Comment (ilovepi) [reply clang/docs/ShadowCallStack.rst:62]: I guess I'm in the minority here in thinking that the compiler should pick based on the HW capabilities.  I think I'd be surprised if I was compiling w/ SCS on HW that supported it and got the software SCS... but at the end of the day, as long as we document the usage clearly its probably fine. I acknowledge the concern about falling back to a less secure method, but it still feels like the wrong tradeoff to me. That said, if there's a consensus that the other way makes more sense(which there seems to be), then I'm 100% fine with that. 

Related: what are our thoughts about a frontend warning/diagnostic when the mcpu supports the feature, but hasn't been selected?

Comment (enh-google) [reply clang/docs/ShadowCallStack.rst:62]: > Related: what are our thoughts about a frontend warning/diagnostic when the mcpu supports the feature, but hasn't been selected?

isn't the problem there that it's unlikely to be a decision that's in the developer's hands? (because the real question is "does the OS support it?", and i think that's the counter-suggestion? "decide based on the whole triple", in effect, since taking Android as an example it's likely to be something like "if android && riscv64 && api level >= where you can assume hardware scs works".)

(apologies if this isn't what's really being talked about here --- i came in very late to a long conversation :-) )

Comment (jrtc27) [reply clang/docs/ShadowCallStack.rst:62]: I think it's really (non-Android) Linux that's the odd one out due to not having a version number in the triple?

Comment (jrtc27) [reply clang/docs/ShadowCallStack.rst:62]: (Though you still have the related problem that some older LLVM may not know that OS version X supports it, and still default it to off)

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:56]: STI.hasStdExtZicfiss()

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:117]: STI.hasStdExtZicfiss()

**Comment by topperc** [comment]
Will https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/417 be implemented in a separate patch?

**Review by topperc — APPROVED** [review]
LGTM. We can resolve the question of default behavior later.

**Review by ilovepi — APPROVED** [review]
LGTM. Seems like all of my comments have been addressed. Thanks for the hard work!

**Pull request merged**
